### PR TITLE
Add domoticz

### PR DESCRIPTION
**Pull request by atlej68**
This change is rather large, and I dont expect you to merge it as it is, my code should probably be reorganized/rewritten.
Consider the code more a demonstration on how we can integrate with smarthouse software without the need for masaging the default json or raw payload outside the esp before sending to smarthouse virtual sensors. I think this type of integration can be useful also for other smarthouse software with virtual sensors.

New: 
-additional option for MQTT payload, that adds domoticz specific messages.  
-new menue for configuring  3 type of domoticz sensors.

### CODE REVIEW

**Comment by gskjold** [comment]
This looks pretty good! I have changed the base to 1.3.0 development branch. Could you resolve the conflicts for this branch?

File: src/AmsToMqttBridge.ino [hunk 1]
```diff
@@ -462,6 +465,154 @@ void readHanPort() {
 						String msg;
 						serializeJson(json, msg);
 						mqtt.publish(config.getMqttPublishTopic(), msg.c_str());
+					// 
+					// Start DOMOTICZ
+					//
+					} else if(config.getMqttPayloadFormat() == 2) {
+						StaticJsonDocument<512> json;
+						hanToJson(json, data, hw, temperature);
```
Comment (gskjold) [inline src/AmsToMqttBridge.ino:473]: Drop converting data to JSON

File: src/AmsToMqttBridge.ino [hunk 2]
```diff
@@ -462,6 +465,154 @@ void readHanPort() {
...
+			
+						int idx1 = config.getDomoELIDX();
+						// TODO, this should be configurable....  
+						if (idx1 > 0) {
+							String PowerEnergy;
+							int p;
+							// double energy = config.getDomoEnergy();
+							double tmp_energy;
+							StaticJsonDocument<200> json_PE;
+							p = json["data"]["P"].as<int>();
```
Comment (gskjold) [inline src/AmsToMqttBridge.ino:514]: Get "P" with data.getActiveImportPower()

File: src/AmsConfiguration.cpp [hunk 3]
```diff
@@ -603,6 +670,128 @@ bool AmsConfiguration::loadConfig81(int address) {
...
+	bool domo = false;
+	address += readBool(address, &domo);
+
+	ackWifiChange();
+
+	return true;
+}
+//
+// domoticz (based on 81)
+//
```
Comment (gskjold) [inline src/AmsConfiguration.cpp:682]: Merge this with loadConfig82 which will be the configuration version for v1.3.0

File: src/AmsConfiguration.cpp [hunk 4]
```diff
@@ -603,6 +670,128 @@ bool AmsConfiguration::loadConfig81(int address) {
 	address += readInt(address, &i);
 	setDebugLevel(i);
 
+	bool domo = false;
+	address += readBool(address, &domo);
+
```
Comment (gskjold) [inline src/AmsConfiguration.cpp:675]: Remove this, earlier config loading should not be modified.

File: src/AmsConfiguration.cpp [hunk 5]
```diff
@@ -661,7 +850,20 @@ bool AmsConfiguration::save() {
...
+//  Domoticz 
+//
+	if(domoELIDX) {
+		address += saveBool(address, true);
+		address += saveInt(address, domoELIDX);	
+		address += saveInt(address, domoVL1IDX);	
+		address += saveInt(address, domoVL2IDX);
+		address += saveInt(address, domoVL3IDX);
+		address += saveInt(address, domoCL1IDX);
+		//address += saveString(address, String(domoEnergy).c_str());
```
Comment (gskjold) [inline src/AmsConfiguration.cpp:863]: If this is a unused configuration option, please remove

File: src/web/AmsWebServer.cpp [hunk 6]
```diff
@@ -443,6 +479,18 @@ void AmsWebServer::dataJson() {
...
+	} else if(mqtt->connected() && config->getMqttPayloadFormat() == 2 && config->getDomoELIDX() > 0) {
+		domoStatus = "success";
+	} else if(mqtt->lastError() == 0) {
+		domoStatus = "warning";
+	} else {
+		domoStatus = "danger";
+	}
+	json["status"]["domo"] = domoStatus;
+
 	serializeJson(json, jsonStr);
```
Comment (gskjold) [inline src/web/AmsWebServer.cpp:494]: Looks to me like this badge only shows if domoticz is configured or not? If so, is it necessary?

File: web/configmeter.html [hunk 7]
```diff
@@ -25,6 +25,9 @@
                     <a class="nav-link" href="/config-web">Web</a>
                 </li>
                 <li class="nav-item">
+                    <a class="nav-link " href="/config-domoticz">Domoticz</a>
+                </li>
+               <li class="nav-item">
                     <a class="nav-link" href="/config-system">System</a>
```
Comment (gskjold) [inline web/configmeter.html:31]: Just thinking, maybe move this link into MQTT configuration and only show if Domoticz is selected?

File: src/AmsToMqttBridge.ino (hunk 1 above)
Comment (atlej68) [reply src/AmsToMqttBridge.ino:473]: Agree.

File: src/AmsToMqttBridge.ino (hunk 2 above)
Comment (atlej68) [reply src/AmsToMqttBridge.ino:514]: Agree.

File: src/AmsConfiguration.cpp (hunk 3 above)
Comment (atlej68) [reply src/AmsConfiguration.cpp:682]: ok!

File: src/AmsConfiguration.cpp (hunk 4 above)
Comment (atlej68) [reply src/AmsConfiguration.cpp:675]: agree, not intensional.

File: src/AmsConfiguration.cpp (hunk 5 above)
Comment (atlej68) [reply src/AmsConfiguration.cpp:863]: yes. leftover from testing.

File: src/web/AmsWebServer.cpp (hunk 6 above)
Comment (atlej68) [reply src/web/AmsWebServer.cpp:494]: Probably not, we can remove it.  I think your idea of moving the domo menue below mqtt and only show  config menue if enabled is good, but I probably need help to implement.

File: web/configmeter.html (hunk 7 above)
Comment (atlej68) [reply web/configmeter.html:31]: agree, see above.  May need help for this one.

Comment (gskjold) [reply web/configmeter.html:31]: OK no problem, commit whatever you can and I'll make final adjustments

**Comment by atlej68** [comment]
I will try to merge with the 1.3.0 branch, most of the changes needed are small.   I am however new to git and  github.  How do I best merge with the new branch without starting over?

**Comment by gskjold** [comment]
Since you are new, first make sure you committed and pushed your changes to your fork.

Here are the basics, and then some resources to help if you get stuck. Also, if you have trouble, I can do the actual merge and then you just sync your repo with mine.

First add my repo as upstream:
```git remote add upstream https://github.com/gskjold/AmsToMqttBridge.git```
NOTE: Substitute URL if you are using SSH

Then, in your local branch, do:
```git merge upstream/dev-v1.3.0```

https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/configuring-a-remote-for-a-fork

https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

File: src/AmsToMqttBridge.ino (hunk 1 above)
Comment (atlej68) [reply src/AmsToMqttBridge.ino:473]: That part is the same as option 0. Left for now.  it has been nice to have the standard json as well when debugging.

File: src/AmsToMqttBridge.ino (hunk 2 above)
Comment (atlej68) [reply src/AmsToMqttBridge.ino]: implemented.  Similar changes  for voltage and current as well.

File: src/AmsConfiguration.cpp (hunk 3 above)
Comment (atlej68) [reply src/AmsConfiguration.cpp]: done

File: src/AmsConfiguration.cpp (hunk 4 above)
Comment (atlej68) [reply src/AmsConfiguration.cpp]: done

File: src/AmsConfiguration.cpp (hunk 5 above)
Comment (atlej68) [reply src/AmsConfiguration.cpp]: done

File: src/web/AmsWebServer.cpp (hunk 6 above)
Comment (atlej68) [reply src/web/AmsWebServer.cpp:532]: Left for now. You can remove if if you want.  Suggest to do it if/when menues are changed.

File: web/configmeter.html (hunk 7 above)
Comment (atlej68) [reply web/configmeter.html:31]: No changes now. I leave it to you to make the adjustments on menues that you think is right.

File: src/AmsToMqttBridge.ino (hunk 1 above)
Comment (gskjold) [reply src/AmsToMqttBridge.ino:482]: OK, I will remove it before release to preserve memory

**Comment by gskjold** [comment]
Thanks! Will merge when I have control over my local changes

**Comment by atlej68** [comment]
The branch should be in sync with dev 1.3.0 now.  fixed most of your comments.   Built the merged version and testing in the meter now.  found one typo so far..

**Pull request merged**